src/render/draw_heatmap.js

```diff
@@ -0,0 +1,137 @@
+    if (painter.isOpaquePass) return;
+
+    const gl = painter.gl;
+
+    painter.setDepthSublayer(0);
+    painter.depthMask(false);
+
+    // Allow circles to be drawn across boundaries, so that
+    // large circles are not clipped to tiles
+    gl.disable(gl.STENCIL_TEST);
+
+    renderToTexture(gl, painter, layer);
+
+    gl.clearColor(0, 0, 0, 0);
+    gl.clear(gl.COLOR_BUFFER_BIT);
+
+    gl.blendFunc(gl.ONE, gl.ONE);
+
+    for (let i = 0; i < coords.length; i++) {
+        const coord = coords[i];
+
+        // skip tiles that have uncovered parents to avoid flickering
+        const parentTile = sourceCache.findLoadedParent(coord, 0, {});
+        if (parentTile) {
+            const parentId = parentTile.coord.id;
+            if (sourceCache._tiles[parentId] && !sourceCache._coveredTiles[parentId]) continue;
+        }
+
+        const tile = sourceCache.getTile(coord);
+        const bucket: ?HeatmapBucket = (tile.getBucket(layer): any);
+        if (!bucket) continue;
+
+        const programConfiguration = bucket.programConfigurations.get(layer.id);
+        const program = painter.useProgram('heatmap', programConfiguration);
+        programConfiguration.setUniforms(gl, program, layer, {zoom: painter.transform.zoom});
+        gl.uniform1f(program.uniforms.u_radius, layer.paint['heatmap-radius']);
+
+        const scale = painter.transform.zoomScale(tile.coord.z - painter.transform.zoom);
+        gl.uniform1f(program.uniforms.u_extrude_scale, EXTENT / tile.tileSize * scale);
+        gl.uniform1f(program.uniforms.u_devicepixelratio, browser.devicePixelRatio);
+
+        gl.uniform1f(program.uniforms.u_intensity, layer.paint['heatmap-intensity']);
+        gl.uniformMatrix4fv(program.uniforms.u_matrix, false, coord.posMatrix);
+
+        program.draw(
+            gl,
+            gl.TRIANGLES,
+            layer.id,
+            bucket.layoutVertexBuffer,
+            bucket.indexBuffer,
+            bucket.segments,
+            programConfiguration);
+    }
+
+    renderTextureToMap(gl, painter, layer);
+}
+
+function renderToTexture(gl, painter, layer) {
+    gl.activeTexture(gl.TEXTURE1);
+
+    gl.viewport(0, 0, painter.width / 4, painter.height / 4);
+
+    let texture = layer.heatmapTexture;
+    let fbo = layer.heatmapFbo;
+
+    if (!texture) {
+        texture = layer.heatmapTexture = gl.createTexture();
+        gl.bindTexture(gl.TEXTURE_2D, texture);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
+        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.LINEAR);
+        gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGB, painter.width / 4, painter.height / 4, 0, gl.RGB,
+            painter.extTextureHalfFloat ? painter.extTextureHalfFloat.HALF_FLOAT_OES : gl.UNSIGNED_BYTE, null);
```

Comment:
What's the purpose of using `HALF_FLOAT_OES` when it's supported?  Is it to increase the precision we use for kde values?

Reply:
Yes. Try disabling it and see how the heatmap looks. Better precision allows for a much smoother look.